File: 01_Arrays_&_Hashing/05_Top_K_Frequent_Elements/0347-top-k-frequent-elements.cpp
/*
Problem: LeetCode 347 - Top K Frequent Elements

Description:
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

Intuition:
To find the k most frequent elements, we can utilize a combination of a hash map and bucket sort. By counting the frequencies of elements using the hash map and using bucket sort to group elements by their frequencies, we can efficiently find the k most frequent elements.

Approach:
1. Initialize an empty hash map, m, to store the frequencies of elements.
2. Iterate through each element num in the input array nums:
    - Increment the frequency count of num in the hash map.
3. Create a vector of vectors, buckets, to act as buckets for grouping elements based on their frequencies.
4. Iterate through the key-value pairs in the hash map:
    - Place each key (element) in the corresponding bucket based on its frequency.
5. Create an empty vector, result, to store the k most frequent elements.
6. Iterate from the highest bucket index down to 0:
    - If the result vector size reaches k, break the loop.
    - If the current bucket is not empty, append its elements to the result vector.
7. Return the result vector, which represents the k most frequent elements.

Time Complexity:
The time complexity of this approach is O(n), where n is the size of the input array nums. This is because we iterate through the array once to count the frequencies and place elements in the buckets.

Space Complexity:
The space complexity is O(n), as we store the frequencies of elements in the hash map and the bucket vectors.
*/

class Solution {
  public:
    vector<int> topKFrequent(vector<int> &nums, int k) {
        int n = nums.size();
        unordered_map<int, int> m;

        for (int i = 0; i < n; i++) {
            m[nums[i]]++;
        }

        vector<vector<int>> buckets(n + 1);

        for (auto it = m.begin(); it != m.end(); it++) {
            buckets[it->second].push_back(it->first);
        }

        vector<int> result;

        for (int i = n; i >= 0; i--) {
            if (result.size() >= k) {
                break;
            }

            if (!buckets[i].empty()) {
                result.insert(result.end(), buckets[i].begin(), buckets[i].end());
            }
        }

        return result;
    }
};

/*
! Using Priority Queue
Problem: LeetCode 347 - Top K Frequent Elements

Description:
Given an integer array nums and an integer k, return the k most frequent elements. You may return the answer in any order.

Intuition:
To find the k most frequent elements, we can utilize a combination of a hash map and a priority queue (min-heap). By counting the frequencies of elements using the hash map and maintaining a min-heap of size k, we can efficiently find the k most frequent elements.

Approach:
1. Initialize an empty hash map to store the frequencies of elements.
2. Iterate through each element num in the input array nums:
    - Increment the frequency count of num in the hash map.
3. Initialize a min-heap to store the k most frequent elements based on their frequencies.
4. Iterate through the elements and frequencies in the hash map:
    - Push the current element into the min-heap.
    - If the size of the min-heap exceeds k, remove the smallest element (with the lowest frequency).
5. Return the elements in the min-heap, which represent the k most frequent elements.

Time Complexity:
The time complexity of this approach is O(n log k), where n is the size of the input array nums.
This is because we iterate through the array once to count the frequencies and perform log k operations for each element insertion and removal in the min-heap.

Space Complexity:
The space complexity is O(n), as we store the frequencies of elements in the hash map and the k most frequent elements in the min-heap.
*/
/*
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> freqMap;
        for (int num : nums) {
            freqMap[num]++;
        }

        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> minHeap;
        for (auto& kvp : freqMap) {
            minHeap.push({kvp.second, kvp.first});
            if (minHeap.size() > k) {
                minHeap.pop();
            }
        }

        vector<int> result;
        while (!minHeap.empty()) {
            result.push_back(minHeap.top().second);
            minHeap.pop();
        }

        return result;
    }
};
*/

// ! O(n log n) solution
/*
class Solution {
public:
    // Approach - First make a frequency map normally, then insert key value pairs(frequency first, value second) in vector and sort in descending order
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> M;
        vector< pair<int, int> > arr;
        vector<int> result;

        for(int i = 0; i < nums.size(); i++) {
            // it reaches end of map if it didn't find the element
            if(M.find(nums[i]) == M.end())
                M[nums[i]] = 1;
            else
                M[nums[i]]++;
        }

        for(auto i : M) {
            arr.push_back( make_pair(i.second, i.first) );
        }

        sort(arr.rbegin(), arr.rend());

        for(int i = 0; i < k; i++) {
            result.push_back(arr[i].second);
        }
        return result;
    }
};
*/

File: 01_Arrays_&_Hashing/05_Top_K_Frequent_Elements/0347-top-k-frequent-elements.rs
/*
Problem: LeetCode 347 - Top K Frequent Elements

Key Idea:
The key idea is to use a combination of a hashmap to count the frequency of each element and a min heap (priority queue) to efficiently maintain the top k frequent elements.

Approach:
1. Create a hashmap to count the frequency of each element in the input array.
2. Create a min heap (priority queue) of size k to store pairs of elements and their frequencies.
3. Iterate through the hashmap:
   - For each element, if the heap size is less than k, insert it into the heap.
   - If the heap size is already k, compare the element's frequency with the smallest frequency in the heap. If the element's frequency is greater, pop the smallest element from the heap and insert the current element.
4. Extract the elements from the heap to obtain the top k frequent elements.

Time Complexity:
O(n log k), where n is the number of elements in the input array and k is the given value. Inserting an element into the heap takes O(log k) time, and we perform this operation for each of the n elements.

Space Complexity:
O(n), as we use a hashmap to store the frequency of elements, which can have up to n distinct keys.
*/

use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};

impl Solution {
    pub fn top_k_frequent(nums: Vec<i32>, k: i32) -> Vec<i32> {
        let mut freq_map: HashMap<i32, i32> = HashMap::new();

        // Count the frequency of each element
        for num in &nums {
            *freq_map.entry(*num).or_insert(0) += 1;
        }

        // Create a min heap of size k to store the top k frequent elements
        let mut min_heap: BinaryHeap<Reverse<(i32, i32)>> = BinaryHeap::new();

        // Insert elements into the heap
        for (num, freq) in freq_map.iter() {
            if min_heap.len() < k as usize {
                min_heap.push(Reverse((*freq, *num)));
            } else if freq > &min_heap.peek().unwrap().0 .0 {
                min_heap.pop();
                min_heap.push(Reverse((*freq, *num)));
            }
        }

        // Extract the elements from the heap
        let result: Vec<i32> = min_heap.into_iter().map(|rev| rev.0 .1).collect();

        result
    }
}
